### PR TITLE
feat: micronaut-hibernate-reactive Kotlin example

### PR DESCRIPTION
**Pull request by romannaglic**

### CODE REVIEW

File: guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/GenreController.kt [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+                    .headers { headers: MutableHttpHeaders ->
+                        headers.location(
+                            location(
+                                genre!!.id!!
+                            )
+                        )
+                    }
+            }
+            .onErrorReturn(PersistenceException::class.java, HttpResponse.noContent())
+    }
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/GenreController.kt]: Move this method above `delete` (as it is in the Java version), so the callouts are in order

File: guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/GenreController.kt [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/GenreController.kt]: Remove blank lines at end of class body

File: guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/GenreRepositoryImpl.kt [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import org.reactivestreams.Publisher
+import reactor.core.publisher.Flux
+import reactor.core.publisher.Mono
+import javax.persistence.PersistenceException
+
+
+@Singleton // <1>
+class GenreRepositoryImpl(
+    val applicationConfiguration: ApplicationConfiguration, // <2>
+    sf: SessionFactory, // <3>
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/GenreRepositoryImpl.kt]: Rename `sf` to `sessionFactory`?  Might help people see what it is when looking at the code for the first time?

File: guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/BaseMysqlTest.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+@MicronautTest // <1>
+@TestInstance(TestInstance.Lifecycle.PER_CLASS) // <2>
+open class BaseMysqlTest: TestPropertyProvider { // <3>
+    var mysqlContainer: GenericContainer<*>? = null
+
+    @Inject
+    open var application: EmbeddedApplication<*>? = null
+
+
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/BaseMysqlTest.kt]: remove extra blank line

File: guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/BaseMysqlTest.kt [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    @Inject
+    @field:Client("/")
+    lateinit var httpClient: HttpClient // <4>
+
+
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/BaseMysqlTest.kt]: remove extra blank line

File: guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/BaseMysqlTest.kt [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        }
+        return "jdbc:mysql://localhost:" + mysqlContainer!!.getMappedPort(3306) + "/mysql"
+    }
+
+    @AfterAll
+    open fun stop() {
+        mysqlContainer!!.close()
+    }
+
+
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/BaseMysqlTest.kt]: remove blank lines at end of class body

File: guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/GenreControllerTest.kt [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    @Test
+    fun supplyAnInvalidOrderTriggersValidationFailure() {
+        val thrown = Assertions.assertThrows(HttpClientResponseException::class.java) {
+            blockingClient!!.exchange<Any, Any>(HttpRequest.GET("/genres/list?order=foo"))
+        }
+        Assertions.assertNotNull(thrown.response)
+        Assertions.assertEquals(HttpStatus.BAD_REQUEST, thrown.status)
+    }
+
+
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/GenreControllerTest.kt]: Remove extra blank line

File: guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/GenreControllerTest.kt [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+import io.micronaut.test.extensions.junit5.annotation.MicronautTest
+import jakarta.inject.Inject
+import org.junit.jupiter.api.Assertions
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.TestInstance
+
+class GenreControllerTest: BaseMysqlTest() { // <1>
+    var blockingClient: BlockingHttpClient? = null
+    @BeforeEach
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/GenreControllerTest.kt:18]: Add blank line between field and method

File: guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/GenreControllerTest.kt [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    private fun entityId(response: HttpResponse<*>): Long? {
+        val path = "/genres/"
+        val value = response.header(HttpHeaders.LOCATION) ?: return null
+        val id = value.substringAfter(path)
+        if (id.isNullOrBlank()) return null
+        return id.toLong()
+    }
+
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/GenreControllerTest.kt]: Remove blank line at end of class body

File: guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/MicronautguideTest.kt [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+import io.micronaut.runtime.EmbeddedApplication
+import jakarta.inject.Inject
+import org.junit.jupiter.api.Assertions
+import org.junit.jupiter.api.Test
+
+internal class MicronautguideTest : BaseMysqlTest() {
+    @Inject
+    override var application: EmbeddedApplication<*>? = null
+    @Test
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/MicronautguideTest.kt:12]: Add blank line between field and method

File: guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/GenreRepositoryImpl.kt [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+) : GenreRepository {
+    private val sessionFactory: Stage.SessionFactory
+    private val VALID_PROPERTY_NAMES = arrayOf("id", "name")
+
+
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/GenreRepositoryImpl.kt]: remove extra blank line

File: guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/GenreRepositoryImpl.kt [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+    override fun saveWithException(name: String): Publisher<Genre?> {
+        return Mono.fromCompletionStage(sessionFactory.withTransaction { session ->
+            val entity = Genre(id = null, name = name)
+            session.persist(entity)
+                .thenApply<Genre> { throw PersistenceException() }
+        })
+    }
+
+
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/GenreRepositoryImpl.kt]: Remove blank lines at end of class body

File: guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/GenreController.kt [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    @Status(HttpStatus.NO_CONTENT) // <12>
+    fun delete(id: Long): HttpResponse<*> {
+        genreRepository.deleteById(id)
+        return HttpResponse.noContent<Any>()
+    }
+
+    @Post("/ex") // <10>
+    @SingleResult
+    open fun saveExceptions(@Valid @Body cmd: GenreSaveCommand): Publisher<MutableHttpResponse<Genre?>> {
+        return Mono.from(genreRepository.saveWithException(cmd!!.name!!))
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/GenreController.kt]: `cmd` cannot be null, so `!!` is unrequired

```
        return Mono.from(genreRepository.saveWithException(cmd.name!!))
```

File: guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/GenreControllerTest.kt [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+            response = blockingClient!!.exchange(request)
+            Assertions.assertEquals(HttpStatus.NO_CONTENT, response.status)
+        }
+    }
+
+    private fun entityId(response: HttpResponse<*>): Long? {
+        val path = "/genres/"
+        val value = response.header(HttpHeaders.LOCATION) ?: return null
+        val id = value.substringAfter(path)
+        if (id.isNullOrBlank()) return null
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/GenreControllerTest.kt]: Cannot be `null`, so could be `id.isBlank()`

File: guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/BaseMysqlTest.kt [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+import io.micronaut.runtime.EmbeddedApplication
+import io.micronaut.test.extensions.junit5.annotation.MicronautTest
+import io.micronaut.test.support.TestPropertyProvider
+import jakarta.inject.Inject
+import org.junit.jupiter.api.AfterAll
+import org.junit.jupiter.api.TestInstance
+import org.testcontainers.containers.GenericContainer
+import org.testcontainers.containers.wait.strategy.Wait
+import org.testcontainers.utility.DockerImageName
+import java.util.*
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/BaseMysqlTest.kt]: unused import

File: guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/domain/Book.kt [hunk 16]
```diff
@@ -0,0 +1,23 @@
+package example.micronaut.domain
+
+import javax.persistence.*
+import javax.validation.constraints.NotNull
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/domain/Book.kt]: unused import

File: guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/domain/Genre.kt [hunk 17]
```diff
@@ -0,0 +1,22 @@
+package example.micronaut.domain
+
+import com.fasterxml.jackson.annotation.JsonIgnore
+import javax.persistence.*
+import javax.validation.constraints.NotNull
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/domain/Genre.kt]: unused import

File: guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/GenreControllerTest.kt [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+
+import example.micronaut.domain.Genre
+import io.micronaut.core.type.Argument
+import io.micronaut.http.HttpHeaders
+import io.micronaut.http.HttpRequest
+import io.micronaut.http.HttpResponse
+import io.micronaut.http.HttpStatus
+import io.micronaut.http.client.BlockingHttpClient
+import io.micronaut.http.client.HttpClient
+import io.micronaut.http.client.annotation.Client
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/GenreControllerTest.kt]: unused

File: guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/GenreControllerTest.kt [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+import io.micronaut.http.client.exceptions.HttpClientResponseException
+import io.micronaut.runtime.EmbeddedApplication
+import io.micronaut.test.extensions.junit5.annotation.MicronautTest
+import jakarta.inject.Inject
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/GenreControllerTest.kt]: Unused

File: guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/GenreControllerTest.kt [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+import org.junit.jupiter.api.Assertions
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.TestInstance
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/GenreControllerTest.kt]: Unused

File: guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/GenreRepository.kt [hunk 21]
```diff
@@ -0,0 +1,15 @@
+package example.micronaut
+
+import example.micronaut.domain.Genre
+import org.reactivestreams.Publisher
+import java.util.*
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/GenreRepository.kt]: Unused import

File: guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/BaseMysqlTest.kt [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    open fun startMySQL() {
+        if (mysqlContainer == null) {
+            mysqlContainer = GenericContainer(DockerImageName.parse("mysql:oracle"))
+                .withExposedPorts(3306)
+                .withEnv("MYSQL_ROOT_PASSWORD", "pass")
+                .waitingFor(Wait.forLogMessage(".*/usr/sbin/mysqld: ready for connections.*\\n", 2))
+        }
+        if (!mysqlContainer!!.isRunning()) {
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/BaseMysqlTest.kt]:
```suggestion
        if (!mysqlContainer!!.isRunning) {
```

File: guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/BaseMysqlTest.kt [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+            mysqlContainer!!.start()
+        }
+    }
+
+    open fun getMySQLDbUri(): String {
+        if (mysqlContainer == null || !mysqlContainer!!.isRunning()) {
```
Comment (timyates) [inline guides/micronaut-hibernate-reactive/kotlin/src/test/kotlin/example/micronaut/BaseMysqlTest.kt]: Property access can be used here

```suggestion
        if (mysqlContainer == null || !mysqlContainer!!.isRunning) {
```

**Review by timyates — CHANGES_REQUESTED** [review]
Awesome!

Found a few style issues and unused imports, but it's looking great!

File: guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/GenreController.kt (hunk 13 above)
Comment (romannaglic) [reply guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/GenreController.kt]: The cmd cannot be null, but a name attribute can be. I changed a name attribute to non-null

Comment (timyates) [reply guides/micronaut-hibernate-reactive/kotlin/src/main/kotlin/example/micronaut/GenreController.kt]: `cmd.name!!` should now be `cmd.name` 😉

**Comment by timyates** [comment]
Nice, I'll run some local tests and get it merged in...

**Comment by timyates** [comment]
I tested this locally, and it all works

Java 11: https://ge.micronaut.io/s/etxnp3boaxcga
Java 17: https://ge.micronaut.io/s/pmlgor5tpogku

Thanks again @romannaglic 😎

**Pull request merged**